### DISCUSSION TITLE
How should I allow code in dynamic compilation to use comments?

### DISCUSSION THREAD

**Discussion by NMSAzulX** [dotnet/roslyn#58023] [Q&A] [answered]
```c#
string code = @"
 public class A {   
    /// <summary>
    ///  comment will raise a error.
    /// </summary>
    public string Name{get;set;}
}";

var result = compiler.Emit(code);
//result.IsSucceed is false
```
> Error: CS1569 写入 XML 文档文件时出错: Value must be positive. (Parameter 'count')

**Comment by RikkiGibson** [comment]
I think we need to know more about what method `compiler.Emit` is, as well as how `compiler` is being created, in order to answer the question.

**Reply by NMSAzulX** [reply]
> I think we need to know more about what method `compiler.Emit` is, as well as how `compiler` is being created, in order to answer the question.

OK. It is `CSharpCompilation`.

**Reply by jasonmalinowski** [reply]
There is no emit method that takes just a string argument like that: https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.compilation.emit?view=roslyn-dotnet-4.0.1

**Reply by jasonmalinowski** [reply]
Oh unless it's this extension method:

https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.csharp.csharpfilesystemextensions.emit?view=roslyn-dotnet-4.0.1

If so, the string isn't the code, it's the file path...?

**Reply by NMSAzulX** [reply]
Hello everyone. The following is the specific dynamic compilation code. Complation has been added to the syntax tree.
```c#
Compilation.AddSyntaxTrees(xxx);
compileResult = Compilation.Emit(
              outputStream,
              pdbStream: pdbStream,
              xmlDocumentationStream: xmlStream,
              options: new EmitOptions(pdbFilePath: OutputPdbPath, debugInformationFormat: DebugInformationFormat.PortablePdb));
```

**Reply by svick** [reply]
@NMSAzulX Can you provide a complete program that reproduces the issue?

**Reply by jasonmalinowski** [reply]
Compilation.AddSyntaxTrees returns a _new_ Compilation with the syntax tree added; you need to hold onto that and Emit *that*.

**Reply by NMSAzulX** [reply]
> @NMSAzulX Can you provide a complete program that reproduces the issue?  

and @jasonmalinowski   Thank you.  Here is my [UT-test code](https://github.com/dotnetcore/Natasha/blob/master/test/NatashaUT/BuilderUT/AssemblyBuilderTest.cs).    
Code that does not use comments can be compiled successfully.

[Here is the implementation of the compilation option.](https://github.com/dotnetcore/Natasha/blob/master/src/Natasha.CSharp/Natasha.CSharp.Engine/Component/Compiler/NatashaCSharpCompiler.cs#L110)
[Here is the process of compilation.](https://github.com/dotnetcore/Natasha/blob/master/src/Natasha.Framework/Compiler/CompilerBase.cs#L167)

**Reply by NMSAzulX** [reply]
@svick @jasonmalinowski @RikkiGibson  Hello Everyone. I think I have found the problem. I used a syntax tree to format API, and this API caused this problem. 
> return oldTree.GetRoot().NormalizeWhitespace().SyntaxTree;

The result returned by the ToString method is correct, but its structure may have changed.

**Comment by NMSAzulX** [comment, marked as answer]
The answer is : `NormalizeWhitespace()` Method exist a bug.
